pkg/agent/openflow/feature.go

```diff
@@ -0,0 +1,254 @@
+//           \             |             /                  \ ------ \                |          /-----/
+//            \            |            /                             \               |         /
+//            +------------------------+                              +-------------------------+
+//            |       pipeline P       |                              |        pipeline Q       |
+//            |       - table A        |                              |        - table D        |
+//            |       - table B        |                              |        - table F        |
+//            |       - table C        |                              |        - table G        |
+//            +------------------------+                              +-------------------------+
+
+// feature is the interface to program a major function in Antrea data path. The following structures implement this interface:
+// - featurePodConnectivity, implementation of connectivity for Pods, activated by default.
+// - featureNetworkPolicy, implementation of K8s NetworkPolicy and Antrea NetworkPolicy, activated by default.
+// - featureService, implementation of K8s Service, activated by default.
+// - featureEgress, implementation of Egress, activation is determined by feature gate Egress.
+// - featureMulticast, implementation of multicast, activation is determined by feature gate Multicast,
+// - featureTraceflow, implementation of Traceflow.
+type feature interface {
+	// getFeatureName returns the name of the feature.
+	getFeatureName() string
+	// getRequiredTables returns a slice of required tables of the feature.
+	getRequiredTables() []*Table
+	// initFlows returns the initial flows of the feature.
+	initFlows() []binding.Flow
+	// replayFlows returns the fixed and cached flows that need to be replayed after OVS is reconnected.
+	replayFlows() []binding.Flow
+}
+
+// Table in FlexiblePipeline is the basic unit to build OVS pipelines. A Table can be referenced by one or more features,
+// but its member struct ofTable will be initialized and realized on OVS only when it is referenced by any activated features.
+type Table struct {
+	name     string
+	stage    binding.StageID // All stages are defined in pkg/ovs/openflow/interfaces.go
+	pipeline binding.PipelineID
+	ofTable  binding.Table
+}
+
+// tableOrderCache is used to save the order of all defined tables located in file pkg/agent/openflow/pipeline.go. The tables
+// have the same pipeline ID are saved in a slice by order of definition. When building a pipeline, the required table list
+// is aggregated from all activated features, and the order of the required tables is decided by the saved order.
+var tableOrderCache map[binding.PipelineID][]*Table
+
+// newTable is used to declare a Table. A table should belong to a stage and pipeline defined in file pkg/ovs/openflow/interfaces.go.
+// Stage is used to group tables which implement similar functions in a pipeline. Stages include:
```

Comment:
I feel we should declare the constants of various stages after the comments, and put description of each stage before its constant. It does not make sense to put descriptions and declarations in different files, and the stages are indeed part of Antrea flow/pipeline design, not generic OVS/openflow abstractions.

The types for stage/pipeline can be in pkg/ovs/openflow, which can be viewed as generic abstractions.

Reply:
Reasonable, in sum:

- Types for stage/pipeline should be defined in pkg/ovs/openflow.
- A stage/pipeline like `ClassifierStage`/`pipelineIP` should be defined in pkg/agent/openflow